R/helpers_broom.R

```diff
@@ -0,0 +1,204 @@
+#' Convert Object's Attributes to a Tibble
+#'
+#' @description `r lifecycle::badge("experimental")`
+#'
+#' A helper function that interrogates an object to see if it has attributes
+#' and, if so, converts them to a (list of) tibbles.  Columns are named after
+#' attributes.  If attributes have different lengths but recycling is possible,
+#' a single `tibble` is returned.  Otherwise, a `list` of tibbles is returned.
+#'
+#' @param x (`CrmPackObject`)\cr object whose attributes will be interrogated.
+#' @param .ignore (`character`)\cr names of attrributes to be ignored.
+#'
+#' @return A [`tibble`] or `list` of [`tibble`]s containg the values of the
+#' object's attributes.
+#'
+#' @keywords internal
+#' @noRd
+h_handle_attributes <- function(x, .ignore = c("names", "class", "description", "row.names")) {
+  a <- attributes(x)
+  valid_names <- setdiff(names(a), .ignore)
+  lapply(
+    valid_names,
+    function(n) {
+      z <- attr(x, n)
+      rv <- NULL
+      # Some Design classes have attributes that are functions or CrmPackClass objects
+      if (!is.function(z)) {
+        if (length(z) == 1) {
+          if (is(z, "CrmPackClass")) {
+            z <- z |> tidy()
+          }
+          rv <- tibble::tibble(X = z)
+        } else {
+          if (length(z) == 0) {
+            rv <- tibble::tibble(X = NA)
+          } else {
+            if (is(z, "CrmPackClass")) {
+              rv <- z |> tidy()
+            } else {
+              rv <- tibble::tibble(X = list(z))
+            }
+          }
+        }
+        names(rv) <- n
+      }
+      rv
+    }
+  ) |>
+    dplyr::bind_cols()
+}
+
+#' Tidy a Single Slot of a CrmPackObject
+#'
+#' @description `r lifecycle::badge("experimental")`
+#'
+#' A helper function that converts a single slot of a `CrmPackObject` to a tibble.
+#' If the slots value is a `list`, each element of the list is tidied individually.
+#'
+#' @param obj (`CrmPackObject`)\cr object to be converted.
+#' @param slot_name (`character`)\cr name of the slot to be tidied.
+#' @param col (`character`)\cr The name of the corresponding column in the tidied
+#' tibble.  Defaults to `slot_name`.
+#' @param attributes (`flag`)\cr shoud the object's attributes, if any, be added
+#'  to the output tibble
+#'
+#' @return A [`tibble`]
+#'
+#' @keywords internal
+#' @importFrom rlang :=
+#' @noRd
+h_tidy_slot <- function(obj, slot_name, col = NULL, attributes = FALSE) {
+  if (is.list(slot(obj, slot_name))) {
+    return(
+      lapply(
+        slot(obj, slot_name),
+        function(x) {
+          if (is.data.frame(x)) {
+            return(x)
+          } else if (is.list(x) && stringr::str_detect(class(x)[1], stringr::fixed("tbl_"))) {
+            # Already tidied to a list
```

Comment:
```suggestion
            # Already tidied to a list.
```
etc. (general we use sentence style for comments, ending with full stop)